Lib/zipimport.py

```diff
@@ -0,0 +1,649 @@
+            if ispackage:
+                # add __path__ to the module *before* the code gets
+                # executed
+                path = _get_module_path(self, fullname)
+                fullpath = f'{self.archive}{path_sep}{path}'
+                mod.__path__ = [fullpath]
+
+            if not hasattr(mod, '__builtins__'):
+                mod.__builtins__ = __builtins__
+            _bootstrap_external._fix_up_module(mod.__dict__, fullname, modpath)
+            exec(code, mod.__dict__)
+        except:
+            del sys.modules[fullname]
+            raise
+
+        try:
+            mod = sys.modules[fullname]
+        except KeyError:
+            raise ImportError(f'Loaded module {fullname!r} not found in sys.modules')
+        _bootstrap._verbose_message('import {} # loaded from Zip {}', fullname, modpath)
+        return mod
+
+
+    def get_resource_reader(self, fullname):
+        '''Return the ResourceReader for a package in a zip file.
+
+        If 'fullname' is a package within the zip file, return the
+        'ResourceReader' object for the package.  Otherwise return None.
+        '''
+        from importlib import resources
+        return resources._zipimport_get_resource_reader(self, fullname)
+
+
+    def __repr__(self):
+        return f'<zipimporter object "{self.archive}{path_sep}{self.prefix}">'
+
+
+# _zip_searchorder defines how we search for a module in the Zip
+# archive: we first search for a package __init__, then for
+# non-package .pyc, and .py entries. The .pyc entries
+# are swapped by initzipimport() if we run in optimized mode. Also,
+# '/' is replaced by path_sep there.
+_zip_searchorder = (
+    (path_sep + '__init__.pyc', True, True),
+    (path_sep + '__init__.py', False, True),
+    ('.pyc', True, False),
+    ('.py', False, False),
+)
+
+# Given a module name, return the potential file path in the
+# archive (without extension).
+def _get_module_path(self, fullname):
+    return self.prefix + fullname.rpartition('.')[2]
+
+# Does this path represent a directory?
+def _is_dir(self, path):
+    # See if this is a "directory". If so, it's eligible to be part
+    # of a namespace package. We test by seeing if the name, with an
+    # appended path separator, exists.
+    dirpath = path + path_sep
+    # If dirpath is present in self._files, we have a directory.
+    return dirpath in self._files
+
+# Return some information about a module.
+def _get_module_info(self, fullname):
+    path = _get_module_path(self, fullname)
+    for suffix, isbytecode, ispackage in _zip_searchorder:
+        fullpath = path + suffix
+        if fullpath in self._files:
+            return ispackage
+    return None
+
+
+# implementation
+
+def _unpack_uint32(data):
+    """Convert 4 bytes in little-endian to an integer."""
+    assert len(data) == 4
+    return int.from_bytes(data, 'little')
+
+def _unpack_uint16(data):
+    """Convert 2 bytes in little-endian to an integer."""
+    assert len(data) == 2
```

Comment:
Nit: Wouldn't be better if we raised `ValueError` here and in `_unpack_uint32()`?

Reply:
In all cases the length of the data is checked before calling `_unpack_uint*()`. The assertion condition is always true.